File: src/clique.f
      subroutine clique(sim,ds,left,rows,cols,alphac,top,bottom,orig)
c
      integer rows
      integer cols
      double precision sim(cols,cols)
      integer ds(rows,cols)
      integer left(rows)
      double precision alphac
      integer top
      integer bottom
      integer orig
c
      integer point
c    
c*************************** one ***************************
c
      top = 1
c
      do 10 i=1,cols-1
      where (sim(i,i+1:cols) < alphac) ds(top,i+1:cols) = 1
      if (all(ds(top,:) .eq. 1)) then
        cycle
      endif
      if (any(ds(top,:) .eq. 1)) then
        left(top) = i
        top = top + 1
      endif
   10 continue
c
      if (top .eq. 1) then
        return
      else
        orig = top - 1
      endif
c
c
c*************************** two **************************
c
      ds(top,left(1)) = 1
      ds(top,left(2)) = 1
c
      ds(top+1,:) = ds(2,:)
      ds(top+2,:) = ds(1,:)
      ds(top+3,:) = max(ds(1,:),ds(2,:))
      ds(top+1,left(1)) = 1
      ds(top+2,left(2)) = 1
c
      bottom = top + 3
      call collap(ds,top,bottom,rows,cols,orig)
c
      point = bottom 
c
      do 21 i=3,orig
        do 22 j=top,bottom
        if (point+2 > rows) then
          orig = -1
          return
        endif
        ds(point+1,:) = ds(j,:)
        ds(point+1,left(i)) = 1
        ds(point+2,:) = max(ds(i,:),ds(j,:))
        point = point + 2
   22   continue
      top = bottom + 1
      bottom = point
c     bottom = bottom * 2
      call collap(ds,top,bottom,rows,cols,orig)
      point = bottom
   21 continue
c
      end


      subroutine collap(ds,top,bottom,rows,cols,orig)
c
      integer rows
      integer cols
      integer ds(rows,cols)
      integer top 
      integer bottom 
      integer orig
c
      integer point
      integer keeper
c
c****************************** one *************************
c
      keeper = 0
      point = bottom 
c
      do 10 i=top,bottom
        do 11 j=top,i-1
        suma = sum(ds(i,:))
        sumb = sum(max(ds(i,:),ds(j,:)))
        if (all(ds(i,:)-ds(j,:) .eq. 0)) then
          goto 10
        else if (suma .eq. sumb) then
          goto 10
        endif
   11   continue
c
        do 13 j=i+1,bottom
        suma = sum(ds(i,:))
        sumb = sum(max(ds(i,:),ds(j,:))) 
        if (all(ds(i,:)-ds(j,:) .eq. 0)) then
          cycle
        else if (suma .eq. sumb) then
          goto 10
        endif
   13   continue
c
      point = point + 1
      ds(point,:) = ds(i,:)
      keeper = keeper + 1
   10 continue
c
c* collap ********************* two ******************************
c
      do 20 i=1,keeper
      ds(orig+i,:) = ds(bottom+i,:)
   20 continue
c
      top = orig + 1
      bottom = orig + keeper 
c
      return
c
      end

File: R/clique.R
clique <- function (dist,alphac,minsize=1,mult=100) 
{
     if (class(dist) != 'dist') 
         stop("The first argument must be of class 'dist'") 
     if (alphac < 0 || alphac > 1) 
         stop("alphac must be [0,1]")
     sim <- 1-as.matrix(dist)
     rows <- mult * nrow(sim)
     cols <- ncol(sim)
     top <- 0
     bottom <- 0
     orig <-0
     ds <- matrix(0,nrow=rows,ncol=cols)
     left <- rep(0,rows)
     tmp <- .Fortran('clique',
         as.double(sim),
         ds=as.integer(ds),
         as.integer(left),
         as.integer(rows),
         as.integer(cols),
         as.double(1-alphac),
         top=as.integer(top),
         bottom=as.integer(bottom),
         orig=as.integer(orig),
         PACKAGE='optpart')
    if (tmp$orig < 0) {
         print('Memory overflow.  Increase parameter mult and try again')
         out <- NULL
    } else {
         musubx <- 1 - matrix(tmp$ds,ncol=cols)[tmp$top:tmp$bottom,]
         test <- apply(musubx,1,sum) >= minsize 
         musubx <- musubx[test,]
         colnames(musubx) <- attr(dist,'Labels')
         member <- list()
         for (i in 1:nrow(musubx)) {
           member[[i]] <- seq(1:ncol(musubx))[musubx[i,]==1]
         }
         out <- list(musubx=musubx,member=member,alphac=alphac,
                     names= attr(dist,'Labels'))
         attr(out,'class') <- 'clique'
         attr(out,'call') <- match.call()
         attr(out,'timestamp') <- date()
    }
    out
}

clique.test <- function (cliq,env,minsize=2,plotit=FALSE)
{
    size <- nrow(cliq$musubx)
    probs <- rep(NA,size)
    for (i in 1:size) {
        if (sum(cliq$musubx[i,]) < minsize) probs[i] <- NA
        else probs[i] <- envrtest(cliq$musubx[i,],env,plotit=plotit)$prob
        if (plotit) readline('hit return to continue')
    }
    if (plotit) {
        plot(sort(probs))
        abline(h=0.05,col=2)
    }
    invisible(probs)
}

summary.clique <- function(object,...)
{
    num <- nrow(object$musubx)
    minsize <- min(apply(object$musubx>0,1,sum))
    maxsize <- max(apply(object$musubx>0,1,sum))
    cat(paste(num,'maximal cliques at alphac = ',object$alphac),"\n")
    cat(paste('minimum size = ',minsize,"\n"))
    cat(paste('maximum size = ',maxsize,"\n"))
}

plot.clique <- function(x, panel='all', ...)
{
    if (panel == 'all' || panel == 1) {
        plot(sort(apply(x$musubx>0,1,sum)),xlab='clique',ylab='size')
        if (panel == 'all')
            readline('hit return to continue : ')
    }
    if (panel == 'all' || panel == 2) {
        plot(sort(apply(x$musubx>0,2,sum)),xlab='plot',ylab='number of cliques')
    }
}

clique.size <- function (cli) 
{
    if (!inherits(cli,'clique'))
        stop("You must pass an argument of class 'clique' from clique()")
    out <- apply(cli$musubx,1,sum)
    out
}

clique.occ <- function (cli) 
{
    if (!inherits(cli,'clique'))
        stop("You must pass an argument of class 'clique' from clique()")
    out <- apply(cli$musubx,2,sum)
    out
}

clique.members <- function (cli,which='ALL') 
{
    if (!inherits(cli,'clique'))
        stop("The first argument must be of class 'clique' from clique()")

    names <- cli$names
    
    for (i in 1:length(cli$member)) {
        x <- names[cli$member[[i]]]
        if (which=='ALL' | which %in% x) {
            cat("\n")
            cat(x)
        }
    }
    cat ("\n")
}

clique.venn <- function (cli,a,b) 
{
    if (!inherits(cli,'clique'))
        stop("The first argument must be of class 'clique' from clique()")
    numcli <- nrow(cli$musubx)
    if (a > numcli  | b > numcli) 
        stop(paste("Clique numbers must be less than",numcli+1))
    inter <- intersect(cli$member[[a]],cli$member[[b]])
    cat("\n")
    cat(setdiff(cli$member[[a]],cli$member[[b]]))
    cat("\n")
    cat(inter)
    cat("\n")
    cat(setdiff(cli$member[[b]],cli$member[[a]]))
    cat("\n")
}

